File: subtitles/file.py
import requests
import json
import time

def read_file(filename, chunk_size=5242880):
    with open(filename, 'rb') as _file:
        while True:
            data = _file.read(chunk_size)
            if not data:
                break
            yield data

def upload_file(api_token, path):
    """
    Upload a file to the AssemblyAI API.

    Args:
        api_token (str): Your API token for AssemblyAI.
        path (str): Path to the local file.

    Returns:
        str: The upload URL.
    """
    headers = {'authorization': api_token}
    response = requests.post('https://api.assemblyai.com/v2/upload',
                             headers=headers,
                             data=read_file(path))

    if response.status_code == 200:
        return response.json()["upload_url"]
    else:
        print(f"Error: {response.status_code} - {response.text}")
        return None

def create_transcript(api_token, audio_url):
    """
    Create a transcript using AssemblyAI API.

    Args:
        api_token (str): Your API token for AssemblyAI.
        audio_url (str): URL of the audio file to be transcribed.

    Returns:
        dict: Completed transcript object.
    """
    # Set the API endpoint for creating a new transcript
    url = "https://api.assemblyai.com/v2/transcript"

    # Set the headers for the request, including the API token and content type
    headers = {
        "authorization": api_token,
        "content-type": "application/json"
    }

    # Set the data for the request, including the URL of the audio file to be transcribed
    data = {
        "audio_url": audio_url
    }

    # Send a POST request to the API to create a new transcript, passing in the headers and data
    response = requests.post(url, json=data, headers=headers)

    # Get the transcript ID from the response JSON data
    transcript_id = response.json()['id']

    # Set the polling endpoint URL by appending the transcript ID to the API endpoint
    polling_endpoint = f"https://api.assemblyai.com/v2/transcript/{transcript_id}"

    # Keep polling the API until the transcription is complete
    while True:
        # Send a GET request to the polling endpoint, passing in the headers
        transcription_result = requests.get(polling_endpoint, headers=headers).json()

        # If the status of the transcription is 'completed', exit the loop
        if transcription_result['status'] == 'completed':
            break

        # If the status of the transcription is 'error', raise a runtime error with the error message
        elif transcription_result['status'] == 'error':
            raise RuntimeError(f"Transcription failed: {transcription_result['error']}")

        # If the status of the transcription is not 'completed' or 'error', wait for 3 seconds and poll again
        else:
            time.sleep(3)

    return transcription_result

def export_subtitles(api_token, transcript_id, subtitle_format):
    """
    Export subtitles using AssemblyAI API.

    Args:
        api_token (str): Your API token for AssemblyAI.
        transcript_id (str): The ID of the transcript you want to export as subtitles.
        subtitle_format (str): The subtitle format to export (either 'srt' or 'vtt').

    Returns:
        str: The filename of the saved subtitle file.
    """
    # Set the API endpoint for exporting subtitles
    url = f"https://api.assemblyai.com/v2/transcript/{transcript_id}/{subtitle_format}"

    # Set the headers for the request, including the API token
    headers = {
        "authorization": api_token
    }

    # Send a GET request to the API to get the subtitle data
    response = requests.get(url, headers=headers)

    # Check if the response status code is successful (200)
    if response.status_code == 200:
        # Save the subtitle data to a local file
        filename = f"{transcript_id}.{subtitle_format}"
        with open(filename, "wb") as subtitle_file:
            subtitle_file.write(response.content)
        return filename
    else:
        raise RuntimeError(f"Subtitle export failed: {response.text}")

your_api_token = "{your_api_token}"

# Upload a local file
filename = "/path/to/foo.wav"
upload_url = upload_file(your_api_token, filename)

# Transcribe it
transcript = create_transcript(your_api_token, upload_url)

# Call the export_subtitles function with your API token, the transcript ID, and the desired subtitle format
subtitle_format = "srt"  # or "vtt" if you want VTT format
subtitle_filename = export_subtitles(your_api_token, transcript['id'], subtitle_format)

# Print the completed transcript object
print(transcript)

File: subtitles/file.rb
require 'net/http'
require 'httparty'

# The URL for the AssemblyAI API endpoint for creating a new transcription job
url = "https://api.assemblyai.com/v2/transcript"

headers = {
    # The authorization header with your AssemblyAI API token
    "authorization" => "{your_api_token}",

    # The content-type header for the request body
    "content-type" => "application/json"
}

data = {
    # The URL of the audio file to be transcribed
    "audio_url" => "https://bit.ly/3yxKEIY"
}

# Parse the API endpoint URL into a URI object
uri = URI.parse(url)

# Create a new Net::HTTP object for making the API request
http = Net::HTTP.new(uri.host, uri.port)

# Use SSL for secure communication
http.use_ssl = true

# Create a new HTTP POST request with the API endpoint URL and headers
request = Net::HTTP::Post.new(uri.request_uri, headers)

# Set the request body to the data hash, converted to JSON format
request.body = data.to_json

# Send the API request and store the response object
response = http.request(request)

# Get the ID of the new transcription job from the response
transcript_id = response.parsed_response["id"]

# Construct the polling endpoint URL for checking the status of the transcription job
polling_endpoint = "https://api.assemblyai.com/v2/transcript/#{transcript_id}"

# Start an infinite loop that will continue until the transcription job is completed or an error occurs
while true
    # Send a GET request to the polling endpoint and store the response
    polling_response = HTTParty.get(polling_endpoint, headers: headers)

    # Parse the response JSON and store it in a hash
    transcription_result = polling_response.parsed_response

    # If the transcription job has completed successfully, break out of the loop
    if transcription_result["status"] == "completed"
        break
    elsif transcription_result["status"] == "error"
        # If an error occurred during the transcription job, raise an exception with the error message
        raise "Transcription failed: #{transcription_result["error"]}"
    else
        # If the transcription job is still in progress, wait for 3 seconds before checking again
        sleep(3)
    end
end